### ISSUE TITLE
Questions regarding the test accuracy on algorithms.

### ISSUE THREAD

**Issue by Yutong-Dai** [TsingZ0/PFL-Non-IID#16]
First and foremost, thanks for the nice work!

I have following questions.

## Datatset split
When checking dataset split method, it seems that the `separate_data` does not correctly split the dataset. To verify this,  run below commands.

```
# generate the dataset
python generate_cifar100.py noniid -- dir

nohup python -u main.py -lbs 64 -nc 20 -jr 0.4 -nb 100 -data Cifar100 -m cnn -algo FedAvg -gr 100 -did 0 -go reproduce --local_steps 5 --local_learning_rate 0.01> Cifar100_FedAvg.out 2>&1 &
```

Then set a break point on [this line](https://github.com/TsingZ0/PFL-Non-IID/blob/6fd94ffa8006c3866e3e34cb97ff20b66f9088fc/system/flcore/servers/serverbase.py#L196) of the `serverbase.py`. And check the total number of testing dataset using `sum(stats[1])`. It returns `15008` on myside,  which means some training data is also set into testing data as Cifar100 only has 10000 testing sample.

## Reproduce results

I fail to use your code to reproduce the results reported in the FedROD's paper https://arxiv.org/pdf/2107.00778.pdf. I only tested `FedAvg`, `pFedMe`. 

Follow the setting used in the FedROD's paper.
Data partition: Dir(0.1), which is the default setting in `generate_cifar100.py` when `dir` is specified.
Other configurations:
0.4 join ratio / 20 clients / 100 rounds / 5 local epochs / 0.01 local learning rate.

The code I used are given below.

```
# generate the dataset
python generate_cifar100.py noniid -- dir

nohup python -u main.py -lbs 64 -nc 20 -jr 0.4 -nb 100 -data Cifar100 -m cnn -algo FedAvg -gr 100 -did 0 -go reproduce --local_steps 5 --local_learning_rate 0.01> Cifar100_FedAvg.out 2>&1 &

nohup python -u main.py -lbs 64 -nc 20 -jr 0.4 -nb 100 -data Cifar100 -m cnn -algo pFedMe -gr 100 -did 1 -go reproduce --local_steps 5 --local_learning_rate 0.01> Cifar100_pFedMe.out 2>&1 &
```

To make fair comparison, I also modified your in the  to perform extra data augmentation following FedROD's paper.

```
    transform_train = transforms.Compose([transforms.RandomCrop(32, padding=4),
                                          transforms.RandomHorizontalFlip(),
                                          transforms.ToTensor(),
                                          transforms.Normalize(mean=[0.507, 0.487, 0.441],
                                                               std=[0.267, 0.256, 0.276])])
    transform_test = transforms.Compose([transforms.ToTensor(),
                                         transforms.Normalize(mean=[0.507, 0.487, 0.441],
                                                              std=[0.267, 0.256, 0.276])])

    trainset = torchvision.datasets.CIFAR100(root='~/data', train=True,
                                             download=True, transform=transform_train)
    testset = torchvision.datasets.CIFAR100(root='~/data', train=False,
                                            download=True, transform=transform_test)
```

What I got from your code were

For `FedAvg`

```
Best global accuracy.
0.2305243759423689

Average time cost per round.
7.7570583581924435
File path: ../results/Cifar100_FedAvg_reproduce_0.h5

Average time cost: 792.34s.
Length:  101
std for best accurancy: 0.0
mean for best accurancy: 0.2305243759423689
All done!
```

which in the paper, the paper is 41.8%. There is a 20% gap.

For `pFedMe`

```
Evaluate personalized model
Average Personalized Test Accurancy: 0.2894

Best personalized results.
0.30951199338296115

Average time cost: 2207.25s.
Length:  101
std for best accurancy: 0.0
mean for best accurancy: 0.30951199338296115
All done!
```

which in the paper, the paper is 38.6%. There was a 8% gap.

I may miss something. Could you please comment on this? Many thanks.

**Comment by TsingZ0** [comment]
Hi, @Yutong-Dai 

> Then set a break point on [this line](https://github.com/TsingZ0/PFL-Non-IID/blob/6fd94ffa8006c3866e3e34cb97ff20b66f9088fc/system/flcore/servers/serverbase.py#L196) of the serverbase.py. And check the total number of testing dataset using sum(stats[1]). It returns 15008 on myside, which means some training data is also set into testing data as Cifar100 only has 10000 testing sample.

This is true since I merge the training set and test set together first and then split them manually. With `train_size = 0.75` in `dataset_utils.py`, we have around 15000 test samples. :-)

**Comment by TsingZ0** [comment]
As for the reproduction problem, the goal of this platform is not to reproduce the results in the papers, since the experimental settings are hard to in line with all these papers. 

For better accuracy, please tune the hyper-parameters. :-)

**Comment by Yutong-Dai** [comment]
> As for the reproduction problem, the goal of this platform is not to reproduce the results in the papers, since the experimental settings are hard to in line with all these papers.
> 
> For better accuracy, please tune the hyper-parameters. :-)

I totally understand your point :-). But the thing is, using the same hyper parameter as the paper, one may not reproduce the exact results, but at least the result should be similar? Otherwise, it either be the case there some hidden tricks are that we are not aware of or the implementation has some bugs ( with my full respects). 

I also implemented some algorithms from scratch.Even using the same hyper-parameters as the paper, I still cannot reproduce the results. But interestingly or you can see weiredly, I found that `FedAvg`'s local models are consistently the best in the PFL setting. Do you have any insights?

**Comment by TsingZ0** [comment]
To reproduce the results, you need to set specific settings for each paper, it is hard to realize on one platform when these algorithms have their specific settings in their papers. 

> I found that FedAvg's local models are consistently the best in the PFL setting. Do you have any insights?

According to [FedRoD](https://openreview.net/pdf?id=I1hQbx10Kxn), *Here, we found that if we instead keep { w_m } and evaluate them in P-FL, they outperform nearly all the existing P-FL algorithms. In other words, personalized models seem to come for free from the local training step of generic FL.*, this phenomenon indeed happens.

**Comment by TsingZ0** [comment]
Or you use other datasets, such as the [HAR (Human Activity Recognition Using Smartphones Data Set)](https://archive.ics.uci.edu/ml/datasets/human+activity+recognition+using+smartphones) and [PAMAP2 (Physical Activity Monitoring Data Set)](http://archive.ics.uci.edu/ml/datasets/pamap2+physical+activity+monitoring) dataset implemented in my another repo [FL-IoT](https://github.com/TsingZ0/FL-IoT), without manual dataset assignment for the Non-IID settings, since they are naturally Non-IID.
